moarchiving/moarchiving3obj.py

```
@@ -66,6 +72,10 @@ def __init__(self, list_of_f_vals=None, reference_point=None, infos=None,
         to compute the hypervolume.
         infos are an optional list of additional information about the points in the archive.
         """
+        if not _HAS_SORTEDCONTAINERS:
```

Why do we need this? We already have a warning and if it doesn't work it will raise an error anyways?

---

Weird things like this can happen then, when half of the things works and half of the things raise an error in a weird place, which could confuse a user (one that wouldn't pay attention to the warning). 

```
> from moarchiving import get_mo_archive
E:\moarchiving\moarchiving\moarchiving_utils.py:6: UserWarning: `sortedcontainers` module not installed, moarchiving for 3 and 4 objectives will not work
  _warnings.warn('`sortedcontainers` module not installed, moarchiving for 3 and 4 objectives will not work')

> moa = get_mo_archive([[1, 2, 3]], n_obj=3)
Traceback (most recent call last):
  File "C:\conda\envs\moarchiving\lib\site-packages\IPython\core\interactiveshell.py", line 3577, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-5c7000fdc4f2>", line 1, in <module>
    moa = get_mo_archive([[1, 2, 3]], n_obj=3)
  File "E:\moarchiving\moarchiving\get_archive.py", line 72, in get_mo_archive
    return MOArchive3obj(list_of_f_vals, reference_point=reference_point, infos=infos,
  File "E:\moarchiving\moarchiving\moarchiving3obj.py", line 88, in __init__
    self.preprocessing()
  File "E:\moarchiving\moarchiving\moarchiving3obj.py", line 477, in preprocessing
    s = t.outer_delimiter_x(p)
  File "E:\moarchiving\moarchiving\moarchiving_utils.py", line 71, in outer_delimiter_x
    i = self.bisect_left(p) - 1
AttributeError: 'MySortedList' object has no attribute 'bisect_left'

> moa = get_mo_archive(n_obj=3)
> moa.add([1, 2, 3])
True
> list(moa)
[[1, 2, 3]]
```

---

But if you think it's not a problem, then we can also have it like this.

---

I see, that's because you used a hack to get around an error when the import fails :-)

To keep the message clear and produce the least code clutter you could raise the error when the `MySortedList` class is initialized like

```python
    def __init__(self, iterable=None, key=lambda node: node.x[1]):
        """ Initialize the sorted list with the given iterable and key function. """
        if SortedKeyList is list:
            raise ...("`MySortedList `requires `sortedcontainers` to be installed")
        super().__init__(iterable=iterable, key=key)
```
On a side note, `MySortedList` is not a name I would expect to find in production code.

---

Another way to prevent the strange variable in `MOArchive3obj` and yet still raise an error would be replacing 
```python
        if not _HAS_SORTEDCONTAINERS:
```
with 
```python
        try: 
            SortedList
        except NameError:

```
yet this part of the code should never be reached anyways, AFIACS.

---

> On a side note, MySortedList is not a name I would expect to find in production code.

Makes sense. How would you propose to name it? It is an extension to `SortedList`, with some additional functionalities from the paper.

---

A suggestion: `ArchiveSortedList`